### PR TITLE
Feat/add implementation for dfget interface

### PR DESCRIPTION
**Pull request by wangforthinker**
### Ⅰ. Describe what this PR did
Add implementation of http protocol of dfget interface.

### Ⅱ. Does this pull request fix one issue?

NONE.

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
add ut.

### CODE REVIEW

File: pkg/protocol/http/http.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	// DefaultClient is default implementation of Client.
+	DefaultClient = &Client{
+		client:    &http.Client{Transport: DefaultClientOpt.transport},
+		transport: DefaultClientOpt.transport,
+	}
+)
+
+const (
+	// http protocol name
+	ProtocolName = "http"
```
Comment (lowzj) [inline pkg/protocol/http/http.go:41]: Should it contain `http` and `https`?

File: pkg/protocol/http/http.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	protocol.RegisterProtocol(ProtocolName, &ClientBuilder{})
+}
+
+// ClientOpt is the argument of NewProtocolClient.
+type ClientOpt struct {
+	transport http.RoundTripper
+}
+
+// Client is an implementation of protocol.Client for http protocol.
+type Client struct {
```
Comment (lowzj) [inline pkg/protocol/http/http.go:54]: create an anonymous instant to check whether the struct implements the interface, like:
```golang
var _ protocol.Client = &Client{}
```

File: pkg/protocol/http/http.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		hd:     hd,
+		client: cli,
+	}
+}
+
+// ClientBuilder is an implementation of protocol.ClientBuilder for http protocol.
+type ClientBuilder struct{}
+
+func (cb *ClientBuilder) NewProtocolClient(clientOpt interface{}) (protocol.Client, error) {
+	opt, ok := clientOpt.(*ClientOpt)
```
Comment (lowzj) [inline pkg/protocol/http/http.go:86]: Does it allow to `nil` clientOpt?

File: pkg/protocol/http/http.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+	if md != nil {
+		h, ok := md.(*Headers)
+		if ok {
+			hd = h
+		}
+	}
+
+	if hd == nil {
+		hd = NewHTTPMetaData().(*Headers)
```
Comment (lowzj) [inline pkg/protocol/http/http.go:72]: Why must create it? I think, it can be `nil`.

File: pkg/protocol/http/resource.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	res, err := rs.doRequest(req)
+	if err != nil {
+		return 0, err
+	}
+
+	defer res.Body.Close()
+	lenStr := res.Header.Get(config.StrContentLength)
+	if len(lenStr) == 0 {
```
Comment (lowzj) [inline pkg/protocol/http/resource.go:78]: use `lenStr == ""` to instead of this?

File: pkg/protocol/http/http.go (hunk 1 above)
Comment (wangforthinker) [reply pkg/protocol/http/http.go]: Add htttps support.

File: pkg/protocol/http/http.go (hunk 2 above)
Comment (wangforthinker) [reply pkg/protocol/http/http.go:61]: Add it.

File: pkg/protocol/http/http.go (hunk 3 above)
Comment (wangforthinker) [reply pkg/protocol/http/http.go:99]: Allow nil clientOpt.

File: pkg/protocol/http/http.go (hunk 4 above)
Comment (wangforthinker) [reply pkg/protocol/http/http.go]: Allow it.

File: pkg/protocol/http/resource.go (hunk 5 above)
Comment (wangforthinker) [reply pkg/protocol/http/resource.go]: Done.

File: pkg/protocol/http/http.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+)
+
+func init() {
+	protocol.RegisterProtocol(ProtocolHTTPName, &ClientBuilder{})
+	protocol.RegisterProtocol(ProtocolHTTPSName, &ClientBuilder{supportHTTPS: true})
+}
+
+// ClientOpt is the argument of NewProtocolClient.
+type ClientOpt struct {
+	transport http.RoundTripper
```
Comment (wangforthinker) [inline pkg/protocol/http/http.go:55]: change to map[string]interface, tell the user the keys which could be set.

**Pull request merged**